modules/metastorage/src/main/java/org/apache/ignite/metastorage/internal/watch/WatchAggregator.java

```diff
@@ -0,0 +1,160 @@
+    private long idCntr;
+
+    /** */
+    public long add(Key key, WatchListener lsnr) {
+        var watch = new Watch(new KeyCriterion.ExactCriterion(key), lsnr);
+        var id = idCntr++;
+        watches.put(id, watch);
+        return id;
+    }
+
+    /** */
+    public long addPrefix(Key key, WatchListener lsnr) {
+        var watch = new Watch(new KeyCriterion.PrefixCriterion(key), lsnr);
+        var id = idCntr++;
+        watches.put(id, watch);
+        return id;
+    }
+
+    /** */
+    public long add(Collection<Key> keys, WatchListener lsnr) {
+        var watch = new Watch(new KeyCriterion.CollectionCriterion(keys), lsnr);
+        var id = idCntr++;
+        watches.put(id, watch);
+        return id;
+    }
+
+    /** */
+    public long add(Key from, Key to, WatchListener lsnr) {
+        var watch = new Watch(new KeyCriterion.RangeCriterion(from, to), lsnr);
+        var id = idCntr++;
+        watches.put(id, watch);
+        return id;
+    }
+
+    /** */
+    public void cancel(long id) {
+        watches.remove(id);
+    }
+
+    /** */
+    public void cancelAll(Collection<Long> ids) {
+        watches.keySet().removeAll(ids);
+    }
+
+    /** */
+    public AggregatedWatch watch(long revision, BiConsumer<Collection<IgniteBiTuple<Key, byte[]>>, Long> saveRevisionAct) {
+
+        synchronized (watches) {
+            // TODO: Remove this assert with normal processing of empty watch
+            assert !watches.isEmpty();
+
+            return new AggregatedWatch(inferGeneralCriteria(), revision, watchListener(saveRevisionAct));
+        }
+    }
+
+    /** */
+    // TODO: We can do it better than infer range always
+    private KeyCriterion inferGeneralCriteria() {
+        return new KeyCriterion.RangeCriterion(
+            watches.values().stream().map(w -> w.keyCriterion().toRange().getKey()).min(Key::compareTo).get(),
+            watches.values().stream().map(w -> w.keyCriterion().toRange().getValue()).max(Key::compareTo).get()
+        );
+    }
+
+    /** */
+    private WatchListener watchListener(BiConsumer<Collection<IgniteBiTuple<Key, byte[]>>, Long> storeRevision) {
+        // TODO: Remove this assert with normal processing of empty watch
+        assert !watches.isEmpty();
+
+        // Copy watches to separate collection, because all changes on the WatchAggregator watches
+        // shouldn't be propagated to listener watches immediately.
+        // WatchAggregator will be redeployed with new watches if needed instead.
+        final LinkedHashMap<Long, Watch> cpWatches = new LinkedHashMap<>(watches);
+
+        return new WatchListener() {
+
+            @Override public boolean onUpdate(@NotNull Iterable<WatchEvent> evts) {
+                var watchIt = cpWatches.entrySet().iterator();
+                final List<Long> toCancel = new ArrayList<>();
+                while (watchIt.hasNext()) {
+                    var entry = watchIt.next();
+                    var watch = entry.getValue();
```

Comment:
Well, I don't think that using `var` here is a good approach. The real type of the variable is not obvious to me, especially without IDE (reviewing changes via GitHub UI).
`var watch = cpWatches.entrySet().iterator().next().getValue();`
Perhaps, we should restrict using such a syntax the follwing case: `var xyz = new Xyz();`